### PR TITLE
Use os name for indexing juju agents in simplestreams

### PR DESCRIPTION
**Pull request by wallyworld**
Update agent binary simplestreams metadata handling to select and process based on OS type (eg ubuntu, windows etc) rather than series.

A lot of the file count is import churn due to now using new upstreams for juju/version (now v2) and juju/description (now v3).
And much of the line count is test changes s/xenial/ubuntu etc

The key aspect of the work is to alter agent binary lookup at bootstrap time to translate from the bootstrap series to the underlying OS matching that series prior to calling FindTools().  Several places would use an agent binary to make copies for all "relevant" series, but this is now not needed because there's only one copy for any given OS. Upgrade logic has been similarly simplified. The overall LOC reduction reflects the simpler logic.

A concession needed to be made to deal with charm MinJujuVersion. The charm/v8 repo uses juju/version (not v2) and so a v1->v2 conversion function is used in the small number of places where the MinJujuVersion is checked. Juju v3 will remove this since charm/v9 will be updated to use juju/version/v2.

Juju releases will publish new os based simplestreams metadata which is needed to support the new lookup logic. An initial proposed metadata update for 2.9-rc6 has already been published to allow the look up logic to be QAed (see below).

Still TODO in follow up PRs:
- upgrade step to convert existing series based agent metadata in mongo to os based
- tools upload handler to translate series based agent metadata so agent binary streaming for migrations correctly writes to the target model
- serialisation of version.Binary in a few API calls needs a compatibility shim
- check Python acceptance test harness used to stream agent metadata to ensure CI tests still pass

## QA steps

Check agent lookup logic using published simplestreams. This will ultimately fail to bootstrap because the rc6 agent still uses series based agents. But the juju client logic used at bootstrap to locate agents is tested.
Edit juju version to be "2.9-rc6" instead of rc7 and make install the binaries.

```
$ juju version
2.9-rc6-ubuntu-amd64
$ jujud version
2.9-rc6-ubuntu-amd64
```

```
$ juju bootstrap lxd test
Creating Juju controller "test" on lxd/localhost
Looking for packaged Juju agent version 2.9-rc6 for amd64
Located Juju agent version 2.9-rc6-ubuntu-amd64 at https://streams.canonical.com/juju/tools/agent/2.9-rc6/juju-2.9-rc6-ubuntu-amd64.tgz
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...
 - juju-6687f2-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
...
ERROR cannot read agent metadata in directory /var/lib/juju/tools/2.9-rc6-focal-amd64: open /var/lib/juju/tools/2.9-rc6-focal-amd64/downloaded-tools.txt: no such file or directory
```

Note above the new simplestreams metadata is used 
`Located Juju agent version 2.9-rc6-ubuntu-amd64 at https://streams.canonical.com/juju/tools/agent/2.9-rc6/juju-2.9-rc6-ubuntu-amd64.tgz`

But because it's the already released rc6 agent and the bootstrap setup is driven by the new logic, the set up of the controller agent is to `var/lib/juju/tools/2.9-rc6-ubuntu-amd64` but the old agent still looks for `var/lib/juju/tools/2.9-rc6-focal-amd64`

Now change back to rc7, make install, and bootstrap with --build-agent

```
$ juju bootstrap lxd --bootstrap-series=bionic test --build-agent
Creating Juju controller "test" on lxd/localhost
Building local Juju agent binary version 2.9-rc7 for amd64
Launching controller instance(s) on localhost/localhost...
 - juju-3f654a-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
...
Bootstrap complete, controller "test" is now available
Controller machines are in the "controller" model
Initial model "default" added
```

Do some deploys ensuring machines with different series are created

```
$ juju add-machine --series=bionic
$ juju deploy ubuntu
$ juju status
$ juju status
Model    Controller  Cloud/Region         Version    SLA          Timestamp
default  test        localhost/localhost  2.9-rc7.1  unsupported  13:50:49+10:00

App     Version  Status  Scale  Charm   Store     Channel  Rev  OS      Message
ubuntu  20.04    active      1  ubuntu  charmhub  stable    19  ubuntu  

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   1        10.160.126.105         

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.160.126.251  juju-a6b2e2-0  bionic      Running
1        started  10.160.126.105  juju-a6b2e2-1  focal       Running
```
Agent binaries are unpacked to directories named after the OS
```
$ juju ssh 0 "ls /var/lib/juju/tools"
2.9-rc7.1-ubuntu-amd64  machine-0

$ juju ssh 1 "ls /var/lib/juju/tools"
2.9-rc7.1-ubuntu-amd64  machine-1  unit-ubuntu-0
```

Upgrade the controller
```
$ juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 2.9-rc7.2 (built from source)
best version:
    2.9-rc7.2
started upgrade to 2.9-rc7.2
$ juju status -m controller
Model       Controller  Cloud/Region         Version    SLA          Timestamp
controller  test        localhost/localhost  2.9-rc7.2  unsupported  14:10:03+10:00

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.160.126.198  juju-3f654a-0  bionic      Running

$ juju ssh -m controller 0 "ls /var/lib/juju/tools"
2.9-rc7.1-ubuntu-amd64  2.9-rc7.2-ubuntu-amd64  machine-0
```
Upgrade the model
```
$ juju upgrade-model
best version:
    2.9-rc7.2
started upgrade to 2.9-rc7.2

$ juju status
Model    Controller  Cloud/Region         Version    SLA          Timestamp
default  test        localhost/localhost  2.9-rc7.2  unsupported  14:11:17+10:00

App     Version  Status  Scale  Charm   Store     Channel  Rev  OS      Message
ubuntu  20.04    active      1  ubuntu  charmhub  stable    19  ubuntu  

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   1        10.160.126.105         

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.160.126.251  juju-a6b2e2-0  bionic      Running
1        started  10.160.126.105  juju-a6b2e2-1  focal       Running

$ juju ssh 0 "ls /var/lib/juju/tools"
2.9-rc7.1-ubuntu-amd64  2.9-rc7.2-ubuntu-amd64  machine-0
```

Bootstrap another controller
```
$ juju bootstrap lxd test2 --bootstrap-series=bionic
$ juju switch controller
$ juju destroy-model default -y
```

Increment the controller version a couple of times to get it higher than the source model
```
$ juju upgrade-controller
$ juju upgrade-controller
```

Migrate "default" model from test -> test2
```
$ juju switch test
$ juju migrate default test2
Migration started with ID "71c9622a-c6d4-46df-8f69-95bc84a6b2e2:0"
$ juju status
$ juju status
Model    Controller  Cloud/Region         Version    SLA          Timestamp       Notes
default  test        localhost/localhost  2.9-rc7.2  unsupported  14:20:17+10:00  migrating: aborted, removing model from target controller: model data transfer failed, failed to migrate binaries: charm ch:amd64/focal/ubuntu-19 unexpectedly assigned ch:focal/ubuntu-19
```

^^^^ the error above is a bug in the work to weave arch into the charm URL. The agent binary changes are not the issue here. 

Commenting out the offending code
```
		if usedCurl, err := config.CharmUploader.UploadCharm(curl, content); err != nil {
			return errors.Annotate(err, "cannot upload charm")
		} else if usedCurl.String() != curl.String() {
			// The target controller shouldn't assign a different charm URL.
			//return errors.Errorf("charm %s unexpectedly assigned %s", curl, usedCurl)
		}
```
Migration works except there's now another charm URL bug

```
$ juju migrate default test2
Migration started with ID "71c9622a-c6d4-46df-8f69-95bc84a6b2e2:3"

$ juju status
Model    Controller  Cloud/Region         Version    SLA          Timestamp       Notes
default  test        localhost/localhost  2.9-rc7.2  unsupported  14:45:56+10:00  migrating: validating, waiting for agents to report back

App     Version  Status  Scale  Charm   Store     Channel  Rev  OS      Message
ubuntu  20.04    active      1  ubuntu  charmhub  stable    19  ubuntu  

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   1        10.160.126.105         

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.160.126.251  juju-a6b2e2-0  bionic      Running
1        started  10.160.126.105  juju-a6b2e2-1  focal       Running

$ juju status
ERROR Model "admin/default" has been migrated to controller "test2".
To access it run 'juju switch test2:admin/default'.

$ juju status -m test2:default
ERROR failed to parse charm: cannot parse name and/or revision in URL "": name "" not valid
Model    Controller  Cloud/Region         Version    SLA          Timestamp
default  test2       localhost/localhost  2.9-rc7.2  unsupported  14:46:12+10:00

App     Version  Status  Scale  Charm  Store  Channel  Rev  OS       Message
ubuntu                       0                           0  unknown  

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.160.126.251  juju-a6b2e2-0  bionic      Running
1        started  10.160.126.105  juju-a6b2e2-1  focal       Running
```

Also check that a 2.8 model can be migrated to a 2.9 controller and then subsequently be upgraded using `upgrade-model`, and that `juju ssh <some machine id> "ls -l /var/lib/juju/tools"` has the symlinks to the new 2.9 ubuntu binaries.
^^^ see the error in status related to charm URL revision

### CODE REVIEW

**Comment by wallyworld** [comment]
[os based agent meadata part 1](https://trello.com/c/JkHFVZOB/1463-os-based-agent-meadata-part-1)

File: api/charms/client.go [hunk 1]
```diff
@@ -146,7 +146,7 @@ func (c *Client) AddCharm(curl *charm.URL, origin apicharm.Origin, force bool) (
 		URL:    curl.String(),
 		Origin: origin.ParamsCharmOrigin(),
 		Force:  force,
-		// Deprecated: Series is used here to communicate with older
+		// Deprecated: Release is used here to communicate with older
```
Comment (SimonRichardson) [inline api/charms/client.go:149]: This is wrong, should still be Series

File: api/charms/client.go [hunk 2]
```diff
@@ -174,7 +174,7 @@ func (c *Client) AddCharmWithAuthorization(curl *charm.URL, origin apicharm.Orig
 		Origin:             origin.ParamsCharmOrigin(),
 		CharmStoreMacaroon: csMac,
 		Force:              force,
-		// Deprecated: Series is used here to communicate with older
+		// Deprecated: Release is used here to communicate with older
```
Comment (SimonRichardson) [inline api/charms/client.go:177]: This is wrong, should still be Series

File: apiserver/common/mocks/upgradeseries.go [hunk 3]
```diff
@@ -107,15 +107,15 @@ func (mr *MockUpgradeSeriesMachineMockRecorder) RemoveUpgradeSeriesLock() *gomoc
 // Series mocks base method
 func (m *MockUpgradeSeriesMachine) Series() string {
 	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "Series")
+	ret := m.ctrl.Call(m, "Release")
```
Comment (SimonRichardson) [inline apiserver/common/mocks/upgradeseries.go:110]: This is wrong, regenerate the mocks.

File: apiserver/facades/schema.json [hunk 4]
```diff
@@ -32959,7 +32963,7 @@
                         "Patch": {
                             "type": "integer"
                         },
-                        "Series": {
+                        "Release": {
```
Comment (SimonRichardson) [inline apiserver/facades/schema.json:32966]: So this is worrying, either it's a Find&Replace issue or it's caused a schema change. If it's a schema change then we should seriously work out why.

**Review by SimonRichardson — CHANGES_REQUESTED** [review]

File: api/charms/client.go (hunk 1 above)
Comment (wallyworld) [reply api/charms/client.go:149]: doh, thanks

File: apiserver/facades/schema.json (hunk 4 above)
Comment (wallyworld) [reply apiserver/facades/schema.json:32966]: I have a TODO in the PR description:
"serialisation of version.Binary in a few API calls needs a compatibility shim"

Sadly we use version.Binary directly in apiserver/params (with no tags in the struct). I need to analyse the use cases and see what needs a compatibility shim added. I didn't want to make this PR any bigger bu doing that work here.

**Review by SimonRichardson — APPROVED** [review]
From the code alone, this looks much better.

Also **massive** +1 on the PR description. Thank-you!

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**